### PR TITLE
[FEAT][kernels]: add backward support for fused logp CUDA kernels

### LINKED ISSUE #100: # [FEAT][kernels]: fused logp CUDA kernels lack backward functions
## Problem

The fused logp CUDA kernels are **forward-only**. Every symbol exported from the extension is a forward variant (`csrc/ops.cpp`):

```
fused_logp, fused_logp_sm90,
fused_logp_forward_{out, fp32, indexed_out, indexed_fp32, online_out, online_fp32, online_indexed_out, online_indexed_fp32}
```
Neither kernel file implements a gradient:
- `csrc/fused_logp_kernel.cu` — generic fused logp (all RL variants)
- `csrc/cuda/fused_logp_sm90.cu` — TMA-accelerated SM90 path

## Why it matters

`logp(logits, token_ids)` is differentiable math: in RL training the policy logits are the live network output, and the loss must backprop through the selected log-probs.

## Proposed solution
```
d logp[a] / d logits[v] = 1[v == a] - p[v]
grad_logits[v] = grad_out * (1[v == a] - p[v])
```

## Affected files

- `csrc/fused_logp_kernel.cu` — add backward kernel (covers all forward variants)
- `csrc/cuda/fused_logp_sm90.cu` — add backward for the TMA path
- `csrc/ops.cpp` — bind the backward
- `rl_engine/kernels/ops/cuda/loss/logp.py` — wrap in `torch.autograd.Function`
- `tests/` — gradient tests vs the native reference

### PR DESCRIPTION
**Pull request by Dnoob**
Closes #100

## Summary

- Add CUDA backward kernels for fused logp, including indexed variants.
- Route the generic two-pass, online, and indexed allocating APIs through `torch.autograd.Function` when gradients are required.
- Make the experimental SM90 TMA path differentiable by returning softmax statistics from its forward pass and reusing the generic CUDA backward kernel.
- Preserve the existing no-grad forward results and output dtype behavior.

## Implementation details

- Save `row_max` and `log_sum` separately to reconstruct probabilities without losing precision under large constant logit shifts.
- Compute `grad_out * (one_hot(token_id) - softmax(logits))` without materializing a logits-sized probability tensor.
- Return exactly zero gradients for rows omitted by indexed variants.
- Keep caller-provided `*_out` and deterministic variants forward-only.
- Add extension bindings, type stubs, documentation, and focused gradient and regression tests.

## SM90 fixes

H100 validation also exposed issues in the existing experimental SM90 forward path:

- Limit the TMA tile width to the supported `CUtensorMap` box dimension.
- Account for the full configured tile in the TMA transaction barrier, including a partial last tile.
- Replace the block-wide reduction with consumer-only warp reductions.
- Fall back to the generic CUDA backend when the bf16 row stride is not 16-byte aligned.

## Validation

The backward test suite was run against the compiled CUDA extension on an NVIDIA H100.

```text
root@b34a45b9a263:/workspace/RL-Kernel# nvidia-smi
Sun Jul 19 09:46:37 2026
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.126.09             Driver Version: 580.126.09     CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:AB:00.0 Off |                    0 |
| N/A   27C    P0             69W /  700W |       0MiB /  81559MiB |      0%      Default |
+-----------------------------------------+------------------------+----------------------+

root@b34a45b9a263:/workspace/RL-Kernel# python -c "import torch; print(f'PyTorch {torch.__version__}, CUDA {torch.version.cuda}, capability {torch.cuda.get_device_capability()}')"
PyTorch 2.4.1+cu124, CUDA 12.4, capability (9, 0)

root@b34a45b9a263:/workspace/RL-Kernel# PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 RL_KERNEL_REQUIRE_EXT=1 CUDA_LAUNCH_BLOCKING=1 python -m pytest tests/test_fused_logp_backward.py
================================================= test session starts ==================================================
platform linux -- Python 3.11.10, pytest-9.1.1, pluggy-1.6.0
rootdir: /workspace/RL-Kernel
configfile: pyproject.toml
collected 45 items

tests/test_fused_logp_backward.py ............................................. [100%]

================================================== 45 passed in 2.52s ==================================================
```

## Summary by CodeRabbit

* **New Features**
  * Added LSE-enabled fused log-probability APIs across standard, indexed, online, and SM90/Hopper variants.
  * Added fused backward APIs that reuse saved LSE statistics.
  * Added automatic gradient support for supported fused operations, including indexed variants and SM90 execution.
* **Bug Fixes**
  * Invalid targets and unselected indexed rows now produce zero gradients.
  * Forward-only output variants now reject gradient-enabled usage.
* **Documentation**
  * Clarified tensor contracts, backend fallback rules, and differentiability behavior.
* **Tests**
  * Added comprehensive CUDA coverage for forward, backward, LSE, routing, and fallback behavior.

### CODE REVIEW

File: csrc/cuda/fused_logp_sm90.cu [hunk 1]
```diff
@@ -106,17 +134,28 @@ __global__ void fused_logp_online_tma_kernel(
...
         if (consumer_tid == 0) {
+            float log_sum = logf(row_sum);
+            if (max_out != nullptr) {
+                max_out[row_idx] = row_max;
+                logsum_out[row_idx] = log_sum;
+            }
             int label_idx = labels[row_idx];
             float label_val = __bfloat162float(logits_gmem[row_idx * vocab_size + label_idx]);
-            output_logp[row_idx] = label_val - row_max - logf(row_sum);
+            output_logp[row_idx] = label_val - row_max - log_sum;
```
[the remark this answers is not in the document]
Comment (Dnoob) [reply csrc/cuda/fused_logp_sm90.cu]: Fixed

**Comment by Flink-ddd** [comment]
please resolve the code conflicts and CI error first. Thanks.

**Comment by Dnoob** [comment]
@Flink-ddd Conflicts resolved and all checks are passing now. Could you take a look when you have time? Thanks!

**Comment by Flink-ddd** [comment]
I missed the message. Could you resolve the conflicts again? Thanks.

**Comment by Flink-ddd** [comment]
cc @KJLdefeated PTAL, Thank you.

**Comment by Dnoob** [comment]
Rebased onto `main`, conflicts resolved, DCO fixed.

Conflict note: `main` added a Python-level `_FusedLogpAutograd` in `logp.py`. This PR provides the same gradient via the fused CUDA backward, so I kept this PR's path and removed `_FusedLogpAutograd`. Gradient tests still pass.

Failing checks are pre-existing on `main`: `linting` (black on `vllm_runtime.py` / `flash_attn.py`, not touched here) and `gpu-tests` (RunPod balance too low). CodeRabbit's `ops.cpp` findings are in upstream bindings this PR doesn't modify.

**Comment by Dnoob** [comment]
@Flink-ddd @KJLdefeated  PTAL. Thanks!
